### PR TITLE
Update parquet to depend on arrow subcrates

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #2594 

# Rationale for this change
 

`time cargo build --release` before

```
________________________________________________________
Executed in  101.75 secs    fish           external
   usr time  715.02 secs    0.00 millis  715.02 secs
   sys time   14.18 secs    2.08 millis   14.18 secs
```

After

```
________________________________________________________
Executed in   71.16 secs    fish           external
   usr time  463.27 secs    0.72 millis  463.27 secs
   sys time   10.95 secs    1.50 millis   10.95 secs
```

The tall pole is now actually the compression codecs and not arrow :tada:

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: parquet/Cargo.toml [hunk 1]
```diff
@@ -30,6 +30,14 @@ edition = "2021"
 rust-version = "1.62"
 
 [dependencies]
+arrow-array = { version = "26.0.0", path = "../arrow-array", default-features = false, optional = true }
```
Comment (tustvold) [inline parquet/Cargo.toml:33]: It is somewhat unfortunate the number of these, perhaps we should provide re-exports to reduce this. On the flip side, parquet is a very complex crate and so perhaps it is just a bit special in needing all the things :sweat_smile:

**Comment by tustvold** [comment]
Need to revisit how we plumb in test utilities :cry:

File: parquet_derive/Cargo.toml [hunk 2]
```diff
@@ -34,5 +34,5 @@ proc-macro = true
 [dependencies]
 proc-macro2 = { version = "1.0", default-features = false }
 quote = { version = "1.0", default-features = false }
-syn = { version = "1.0", default-features = false }
-parquet = { path = "../parquet", version = "26.0.0" }
+syn = { version = "1.0", features = ["extra-traits"] }
+parquet = { path = "../parquet", version = "26.0.0", default-features = false }
```
Comment (tustvold) [inline parquet_derive/Cargo.toml:38]: This is part drive-by-cleanup, part fix as it was relying on `multiversion` which is a transitive dependency of `arrow` to enable the `syn` features it requires. As parquet no longer depends on arrow, this caused issues

File: parquet/src/errors.rs [hunk 3]
```diff
@@ -19,8 +19,8 @@
 
 use std::{cell, io, result, str};
 
-#[cfg(any(feature = "arrow", test))]
```
Comment (tustvold) [inline parquet/src/errors.rs:22]: This is necessary to avoid having to declare all the various subcrates as dev-dependencies

File: parquet/Cargo.toml [hunk 4]
```diff
@@ -70,9 +78,9 @@ all-features = true
 [features]
 default = ["arrow", "snap", "brotli", "flate2", "lz4", "zstd", "base64"]
 # Enable arrow reader/writer APIs
-arrow = ["dep:arrow", "base64"]
+arrow = ["base64", "arrow-array", "arrow-buffer", "arrow-cast", "arrow-data", "arrow-schema", "arrow-select", "arrow-ipc"]
```
Comment (viirya) [inline parquet/Cargo.toml:81]: Seems they can combined?

Comment (tustvold) [reply parquet/Cargo.toml:81]: What do you mean?

File: parquet/src/column/reader.rs [hunk 5]
```diff
@@ -515,6 +515,7 @@ where
     /// If the current page is fully decoded, this will NOT load the next page
     /// into the buffer
     #[inline]
+    #[cfg(feature = "arrow")]
     pub(crate) fn peek_next(&mut self) -> Result<bool> {
```
Comment (viirya) [inline parquet/src/column/reader.rs:519]: Hmm, is `peek_next` optional? It is used by some APIs like `skip_records`. If `arrow` is not enabled, what would happen?

Comment (tustvold) [reply parquet/src/column/reader.rs:519]: These changes were necessary to placate clippy as they aren't currently used by anything outside the arrow module

File: parquet/Cargo.toml (hunk 4 above)
Comment (viirya) [reply parquet/Cargo.toml:81]: Nvm, I thought if this `arrow` is unnecessary and we can just have these "base64" features. But I saw you use `arrow` as a whole feature in many places.

File: parquet/src/errors.rs (hunk 3 above)
Comment (viirya) [reply parquet/src/errors.rs:22]: You mean removing `test`?

Comment (tustvold) [reply parquet/src/errors.rs:22]: Yeah, this only worked because arrow is a dev-dependency

Comment (viirya) [reply parquet/src/errors.rs:22]: Hmm, I tried to restore only this line locally, and looks like the tests are okay to run?

**Review by viirya — APPROVED** [review]
Looks good to me.

File: parquet/src/errors.rs (hunk 3 above)
Comment (tustvold) [reply parquet/src/errors.rs:22]: Did you do `cargo test --no-default-features`?

Comment (viirya) [reply parquet/src/errors.rs:22]: Ah, yea, `arrow` is default feature. I didn't exclude it.

File: parquet/Cargo.toml (hunk 1 above)
Comment (viirya) [reply parquet/Cargo.toml:33]: Maybe for some basic crates like arrow-buffer, arrow-data, arrow-schema, perhaps we can provide re-export (arrow-core?) for them.

Like you said, if this is just a special case, then it is fine.
